gstreamer/src/tags.rs

```diff
@@ -264,19 +264,15 @@ impl TagListRef {
     }
 
     pub fn get_index<'a, T: Tag<'a>>(&'a self, idx: u32) -> Option<&'a TypedValue<T::TagType>> {
-        unsafe {
-            let value = ffi::gst_tag_list_get_value_index(
-                self.as_ptr(),
-                T::tag_name().to_glib_none().0,
-                idx,
-            );
-
-            if value.is_null() || (*value).g_type != T::TagType::static_type().to_glib() {
-                return None;
-            }
-
-            Some(&*(value as *const TypedValue<T::TagType>))
-        }
+        self.get_index_generic(T::tag_name(), idx)
+            .and_then(|value| unsafe {
+                let value = value.to_glib_none().0;
+                if (*value).g_type != T::TagType::static_type().to_glib() {
+                    return None;
+                }
+
+                Some(&*(value as *const TypedValue<T::TagType>))
```

Comment:
Why is the `unsafe` here needed? It could a) simply assert if types don't match and there's b) [a way for downcasting](https://github.com/gtk-rs/glib/blob/b8b98c9648326b399f5059f89746d440b8d6f05d/src/value.rs#L641-L643) with safe code that basically does the same you are doing here :)

But otherwise yes, that's what I was thinking of. And the same could possibly be done for the un-indexed getter and the setter/adder function?

Reply:
>  b) a way for downcasting with safe code that basically does the same you are doing here :)

I struggled trying to get that working, but had issues with `downcast` not returning a reference and the `Tag`s types not implementing `FromValueOptional` (can't be more precise right now, sorry). I will try again tonight or later this week.

Reply:
Ok thanks, put your non-working code in a branch then and I'll take a look :)

Reply:
Here's a commit in a dedicated branch with the failing code: https://github.com/fengalin/gstreamer-rs/tree/factorize_tags_fn

AFAICT, `downcast` consumes the `SendValue` and returns an owned `TypedValue` when we get an `Option<&'a SendValue>` and want an `Option<&'a TypedValue<T::TagType>>`.

Reply:
Oh I see. I guess we want a `downcast_ref` for all those variants in `value.rs` in glib :)

Reply:
Let me just add that :)

Reply:
https://github.com/gtk-rs/glib/pull/356

Reply:
Awesome! Thanks! I'm proceeding with the changes in `tags` then.

Reply:
Just one thing, it seems to me there is a mismatch between the comment "Returns `Ok(T)` if the object is an instance of `T` and `Err(self)` otherwise" and the actual signature.

Reply:
Thanks, fixed